Create FungibleTokenMetadataViews contract

alilloig opened the pull request:
Closes: #82 
## Description

Creates the contract for the Fungible Token's Metadata Views.

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-ft/blob/master/CONTRIBUTING.md#styleguides).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels
- [x] Update the version in package.json when there is a change in the smart contracts

alilloig commented:
There are at least two thing I'd love to have some input, the contract name and how to deal with the `MetadataViews.cdc` file

gus-labs commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,177 @@
+
+        /// Function that allows creation of an empty FT vault that is intended
+        /// to store the funds.
+        pub let createEmptyVault: ((): @FungibleToken.Vault)
```
What is the reasoning for this `createEmptyVault` function to be inside the `FTVaultData` struct?

joshuahannan replied:
People can get metadata views from an object without importing or knowing what contract the object comes from, so this view allows code that has the object or a reference to it, but not the imported contract, to be able to create a new vault and also know what the storage and public paths are for it without having to find out that information elsewhere.

joshuahannan reviewed (commented):
Looks good so far! I think the name is good and just copying `MetadataViews` into the utility contracts directory here is probably fine until we figure out a better way to manage it.

Now we just need to add some tests and we can start sharing this PR around a bit more for reviews

satyamakgec commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,177 @@
+        /// that can be displayed in lists, link previews, etc.
+        pub let thumbnail: AnyStruct{MetadataViews.File}
+
+        /// External link to a URL to view more information about this collection.
```
```suggestion
        /// External link to a URL to view more information about the fungible token.
```

satyamakgec commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,177 @@
+        /// Square-sized image to represent this collection.
+        pub let squareImage: MetadataViews.Media
+
+        /// Banner-sized image for this collection, recommended to have a size near 1200x630.
```
```suggestion
        /// Banner-sized image for the fungible token recommended having a size near 1200x630.
```

satyamakgec commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,177 @@
+        /// Banner-sized image for this collection, recommended to have a size near 1200x630.
+        pub let bannerImage: MetadataViews.Media
+
+        /// Social links to reach this collection's social homepages.
```
```suggestion
        /// Social links to reach the fungible token's social homepages.
```

satyamakgec commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,177 @@
+            bannerImage: MetadataViews.Media,
+            socials: {String: MetadataViews.ExternalURL}
+        ) {
+            self.name = name
```
What about the token symbol?

alilloig replied:
I wanna discuss over the FLIP on the forum about the fields of this `FTDisplay`, the symbol is a really good catch, and I also would like to discuss which images we really do need for the FTs, I'm not sure at all that we need a square and a banner Image

satyamakgec replied:
can you please mention the flip here

boczeratul commented:
I love where this is heading! 

From Blocto's perspective, it would be useful to include 2 different kinds of logos in `FTDisplay`:
- One as png/jpg image, which is the colored version of the logo
- One as vector like svg. Since FT logos can appear in multiple places and may need to be changed to different colors depending on the context it's used

alilloig commented:
Hi @boczeratul! It's such an honour that you like this proposal.

What do you think if we turn `pub let logo: MetadataViews.Media` into `pub let logo: MetadataViews.Medias` so we can store any amount of different logos formats?

Just for context

```cadence
    pub struct Medias {

        /// An arbitrary-sized list for any number of Media items
        pub let items: [Media]

        init(_ items: [Media]) {
            self.items = items
        }
    }
```

boczeratul commented:
Awesome 👍

alilloig commented on contracts/FungibleToken.cdc:
```diff
@@ -193,6 +195,24 @@ pub contract interface FungibleToken {
+        /// @return An array of Types defining the implemented views. This value will be used by
+        ///         developers to know which parameter to pass to the resolveView() method.
+        ///
+        pub fun getViews(): [Type]{
```
@joshuahannan I've done the default implementations of the Resolver functions on this branch since it seemed to me that it belonged more to the features than to the example, does it look good?

joshuahannan replied:
yeah, that looks good. The version of Cadence that you're using to test it supports default implementations and is working, right?

alilloig replied:
Not at all, I'm on v0.39.3 of the cli, the latest distributed via brew. I thought the version supporting default implementations was not live yet. I haven't merge this changes into the example token branch which is the one that includes the tests.

<img width="1920" alt="Screenshot 2022-09-27 at 14 17 43 (2)" src="https://user-images.githubusercontent.com/6607145/192524447-ebc04f1f-0ab3-42b9-b52c-46bd20e9f69e.png">

alilloig replied:
Working with v0.41.0 🎉 
<img width="903" alt="Screenshot 2022-10-06 at 18 49 37" src="https://user-images.githubusercontent.com/6607145/194372265-4db040f7-ed2f-4e78-8f33-959b545e11d1.png">

psiemens commented on contracts/FungibleToken.cdc:
```diff
@@ -193,6 +195,24 @@ pub contract interface FungibleToken {
             }
         }
+
+        /// Function that returns all the Metadata Views implemented by a Fungible Token
+        ///
+        /// @return An array of Types defining the implemented views. This value will be used by
+        ///         developers to know which parameter to pass to the resolveView() method.
+        ///
+        pub fun getViews(): [Type]{
```
```suggestion
        /// Function that returns all the Metadata Views implemented by a Fungible Token
        ///
        /// @return An array of Types defining the implemented views. This value will be used by
        ///         developers to know which parameter to pass to the resolveView() method.
        ///
        pub fun getViews(): [Type] {
```

psiemens commented on contracts/FungibleToken.cdc:
```diff
@@ -193,6 +195,24 @@ pub contract interface FungibleToken {
+            return []
+        }
+
+        /// Function that returns a Metadata View out of a Fungible Token
```
```suggestion
        /// Function that resolves a metadata view for this fungible token by type.
```

Nitpick, but IMO I don't think we should treat "Metadata View" and "Fungible Token" as capitalized proper nouns in comments and documentation. I don't feel too strongly about this, though.

alilloig replied:
I'm never sure with this details about use of English, so I'll agree with you!!

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    /// Helper to get a FT view 
```
```suggestion
    /// Helper to get a FT view.
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+        }
+        return FTView(
+            ftDisplay: self.getFTDisplay(viewResolver),
+            vaultData : self.getFTVaultData(viewResolver)
```
```suggestion
            vaultData: self.getFTVaultData(viewResolver)
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+    /// @return A FTView struct
+    ///
+    pub fun getFTView(viewResolver: &{MetadataViews.Resolver}): FTView {
+        let ftView = viewResolver.resolveView(Type<FTView>())
+        if ftView != nil {
+            return ftView! as! FTView
+        }
```
```suggestion
        let maybeFTView = viewResolver.resolveView(Type<FTView>())
        if let ftView = maybeFTView {
            return ftView as! FTView
        }
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+        /// Symbol that could be used as a shorter name for the FT.
+        pub let symbol: String
+
+        /// Description that should be used to give an overview of this FT.
+        pub let description: String
+
+        /// External link to a URL to view more information about the fungible token.
```
```suggestion
        /// Description that should be used to give an overview of the FT.
        pub let description: String

        /// External link to a URL to view more information about the FT.
```

Consistency nitpick :)

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+        /// External link to a URL to view more information about the fungible token.
+        pub let externalURL: MetadataViews.ExternalURL
+
+        /// Image to represent the fungible token logo.
```
```suggestion
        /// One or more versions of the fungible token logo.
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    /// Helper to get FTDisplay in a way that will return a typed optional
```
```suggestion
    /// Helper to get FTDisplay in a way that will return a typed optional.
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+    /// @return A optional FTDisplay struct
+    ///
+    pub fun getFTDisplay(_ viewResolver: &{MetadataViews.Resolver}): FTDisplay? {
+        if let view = viewResolver.resolveView(Type<FTDisplay>()) {
+            if let v = view as? FTDisplay {
+                return v
+            }
+        }
```
```suggestion
        if let maybeDisplayView = viewResolver.resolveView(Type<FTDisplay>()) {
            if let displayView = maybeDisplayView as? FTDisplay {
                return displayView
            }
        }
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    /// Helper to get FTVaultData in a way that will return a typed Optional
```
```suggestion
    /// Helper to get FTVaultData in a way that will return a typed Optional.
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+    /// Helper to get FTDisplay in a way that will return a typed optional
+    /// 
+    /// @param viewResolver: A reference to the resolver resource
+    /// @return A optional FTDisplay struct
```
```suggestion
    /// @return An optional FTDisplay struct
```

psiemens commented on contracts/utilityContracts/MetadataViews.cdc:
```diff
@@ -0,0 +1,740 @@
+import FungibleToken from "./../FungibleToken.cdc"
```
Another nitpick, but I think the `flow-nft` repo stores utility contracts in `contracts/utility`. Feels a bit redundant to mention "contracts" again in `contracts/utilityContracts`

alilloig replied:
You have a pretty good point there, should we change the folder's name here and on the `flow-nft` repo @joshuahannan ???

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,177 @@
+    pub struct FTView {
+        pub let ftDisplay: FTDisplay?     
+        pub let vaultData: FTVaultData?
+        init(
```
Yeah, missing a line break here.

psiemens commented on contracts/FungibleToken.cdc:
```diff
@@ -40,6 +40,8 @@ the deposit function on another user's Vault to complete the transfer.
 
 */
 
+import MetadataViews from "./utilityContracts/Metadataviews.cdc"
```
```suggestion
import MetadataViews from "./utilityContracts/MetadataViews.cdc"
```

psiemens commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,175 @@
+    /// graphics of the FT.
+    ///
+    pub struct FTDisplay {
+        /// Name that should be used when displaying this FT.
+        pub let name: String
+
+        /// Symbol that could be used as a shorter name for the FT.
+        pub let symbol: String
+
+        /// Description that should be used to give an overview of this FT.
+        pub let description: String
```
It'd be nice to include examples for all the fields. For example:

```suggestion
        /// The display name for this token.
        ///
        /// Example: "Flow"
        ///
        pub let name: String

        /// The abbreviated symbol for this token.
        ///
        /// Example: "FLOW"
        ///
        pub let symbol: String

        /// A description the provides an overview of this token.
        ///
        /// Example: "The FLOW token is the native currency of the Flow network."
        ///
        pub let description: String
```

alilloig replied:
Yeah that makes total sense

psiemens reviewed (approved):
The logic looks good! My comments are almost entirely stylistic.

alilloig commented:
Thanks a lot for your comments @psiemens! I love descriptive names on variables so I love the "maybeView" names, it makes the code way more readable!!

psiemens reviewed (approved):
This LGTM!

alilloig commented:
@satyamakgec `package.json` now stands...
```
  "name": "@onflow/flow-ft",
  "version": "1.0.0",
```
According to the PR Template I should move the version to `2.0.0` since we are adding the new contract, am I right?

bjartek commented on contracts/FungibleTokenMetadataViews.cdc:
```diff
@@ -0,0 +1,185 @@
+        ) {
+            pre {
+                receiverLinkedType.isSubtype(of: Type<&{FungibleToken.Receiver}>()): "Receiver public type must include FungibleToken.Receiver."
+                metadataLinkedType.isSubtype(of: Type<&{FungibleToken.Balance, MetadataViews.Resolver}>()): "Metadata public type must include FungibleToken.Balance and MetadataViews.Resolver interfaces."
```
should this also ensure that it is a subtype of receiver? personally I never understood why FT capabilities do not link both receiver and balance.

alilloig replied:
That's a good question. I guess that if we are making sure that a `{FungibleToken.Receiver}` is provided on the `receiverLinkedType` we don't need to check for it on the `metadataLinkedType`.

Also you make me thing if we really want to be checking that the `providerLinkedType` is also a `{MetadataViews.Resolver}`

bjartek reviewed (commented):
LGTM, a small comment about the required restrictions of the balancer cap.

sisyphusSmiling reviewed (approved):
LGTM! Just needs Go assets updated for CI to build properly

joshuahannan commented:
@bshahid331 can you elaborate on what you would want with a code sample? IMO these use cases are not that common and will be phased out eventually, so I don't see much reason to build them into the standard. It might make more sense for those forwarder contracts to upgrade their contracts and relink their paths so that their forwarders expose the metadata though, if that is what you want

joshuahannan commented:
I have one main question. Currently in the FungibleToken contract, we have the Vault resource implementing the `MetadataViews.Resolver` now. This would be a breaking change to all fungible token contracts. Are we ok with that? I feel like we won't want to do any breaking changes until we do the main upgrade to the token standards as part of stable cadence. 

Therefore, if we don't want to introduce a breaking change, then the only other way to include this before then would be to put the metadata views methods into one of the existing interfaces, either `Receiver` or `Balance`. `Balance` makes the most sense to me since it is already kind of a metadata field, but it is unfortunate that the name balance is not accurate to what it fully represents. 

Does anyone have any thoughts? I'll probably just add it to `Balance` if nobody says anything.

alilloig commented:
Last (I hope) push before merging!

Some notes:

- The CI was tied to the secure cadence version, I believe we should be running the CI agains the latest version of the CLI so I removed any specific version flag.
- I created one extra test I just realise I didn't have and split the test suites so they are easier to read, @sisyphusSmiling or @satyamakgec when you have a couple of minutes check the `metadata.test.js` file and let me know it the last new test looks good.
- Anyone can think of a reason why the provider type linked in the `FTVaultData` should include `{MetadataViews.Resolver}` along with `{FungibleToken.Provider}`? I removed that requirement and only required the Resolver interface in the `metadataLinkedType` field. I believe @bjartek had some thoughts about this.
- Finally this will be packing default implementations of the `getViews(): [Type]` and `resolveView(_ view: Type): AnyStruct?` functions inside the `{FungibleToken.Balance}` if nobody is against it.

Thanks to everyone, cheers!!

alilloig commented:
I took the opportunity and fixed a small bug on the switchboard and updated the docs to get everything merged at the same time. If nobody is agains it I'll merge it by Friday and hopefully we can have it deployed to testnet along with the Switchboard next week.
